### ISSUE TITLE
Connection pooling

### ISSUE THREAD

**Issue by george-slastnoy** [microsoft/msphpsql#1044]
.+Please check the [FAQ (frequently-asked questions)](https://github.com/Microsoft/msphpsql/wiki/FAQ) first. If you have other questions or something to report, please address the following:

 +## PHP Driver version or file name
 + 5.6.1
 +## SQL Server version
 + 2008 R2
 +## Client operating system
 + Gentoo Linux
 +## PHP version
 + 7.3.9
 +## Microsoft ODBC Driver version
 + 17.4
+ ## UnixODBC version
+ 2.3.5

Hello, everyone. 

I have enabled connection pooling in odbcinst.ini, but cannot undestand, how connections should be properly returned back to the pool. As I navigate my site, I often see that, instead of reusing an existing connection, a new connection is created. Overtime the number of connections increases up to 10.

My odbcinst.ini is:
[ODBC]
Pooling=Yes

[ODBC Driver 17 for SQL Server]
Description=Microsoft ODBC Driver 17 for SQL Server
Driver=/opt/microsoft/msodbcsql17/lib64/libmsodbcsql-17.4.so.1.1
UsageCount=1
CPTimeout=120

Should I expressly call sqlsrv_close to release a connection back to the pool? Are connections released automatically after redirecting, followed by the exit command?

How can I set min and max pool size? Is this supported in UnixODBC? 

Also, I can connect to Sql server only by specifying a direction IP address in the server part of the connection string. If I specify a DNS, listed in odbc.ini, connect fails.

**Comment by yitam** [comment]
hi @george-slastnoy , note that Gentoo Linux is not listed as one of [the officially supported Linux distributions](https://docs.microsoft.com/sql/connect/odbc/linux-mac/system-requirements?view=sql-server-ver15). 

That being said, [connection pooling in Linux](http://www.unixodbc.org/doc/conn_pool.html) is managed by the Driver Manager. I'd suggest you upgrade [unixODBC to 2.3.7](http://www.unixodbc.org/) because this latest version has fixed some connection pooling issues. To upgrade, please check if [this guide](https://docs.microsoft.com/sql/connect/odbc/linux-mac/installing-the-driver-manager?view=sql-server-ver15 ) helps.

About DNS, please check [this page](https://docs.microsoft.com/sql/connect/odbc/linux-mac/data-access-tracing-with-the-odbc-driver-on-linux?view=sql-server-ver15)

**Comment by george-slastnoy** [comment]
Thank you for advice.
I have updated ODBC to version 2.3.7, enabled tracing, but the problem with opening new connections persists. Also, as I have noticed, pooled connections are not closed according to the timeout value, specified in odbcinst.ini. I have enabled tracing in odbcinst.ini (Trace=Yes, TraceFile=), but sqlsrv driver calls are not logged there. Connections via isql are logged. Also, with the sqlsrv driver I can connect by IP only, not by DSN. Probably there is something wrong with my connection string.

**Comment by george-slastnoy** [comment]
I have updated odbc to 2.3.7, but that did not help. 
I have checked, the driver is OK. Is this up to the driver to decide when to establish a new connection? For some reason connection timeout specified in odbcinst.ini does not work. Connections remain open after the timeout elapses.

In the log my connection is:  Str In = [Server={server_ip};UID={sa};PWD=****************;Database={...};APP={...};MARS_Connection={Yes};Driver={ODBC Driver...][length = 148 (SQL_NTS)]. Is this OKey?
 UNICODE Using encoding ASCII 'UTF8' and UNICODE 'UTF16LE'

**Comment by george-slastnoy** [comment]
I have just read at https://www.easysoft.com/products/data_access/odbc_odbc_bridge/performance_white_paper.html#2_3 , that this might be due to Apache's configuration. Currenlty my Apache has the prefork mode. Should I change it to fix the problem?

**Comment by yitam** [comment]
@george-slastnoy connection pooling is managed by the Driver Manager, unixodbc.

**Comment by george-slastnoy** [comment]
It has turned out, that in order for connection pooling to function properly, it is necessary to run Apache and PHP in threaded (worker) mode. SQLSRV driver needs to be reinstalled (with Pecl, uninstalled and then installed again).

**Comment by dadamssg** [comment]
anyone know to to "reload" the drivers? 

![image](https://user-images.githubusercontent.com/881986/81490772-c3004a00-924b-11ea-81bf-82cace21aca5.png)

I'm using fpm and restarting fpm  still shows pooling as enabled even though i have `Pooling=No` in my `/etc/odbcinst.ini` file. 

![image](https://user-images.githubusercontent.com/881986/81490788-f04cf800-924b-11ea-8a4c-b8f1f251b7cd.png)

**Comment by yitam** [comment]
@dadamssg please note pdo_odbc is not our driver. To check if pooling is enabled or not, use [@@SPID](https://docs.microsoft.com/sql/t-sql/functions/spid-transact-sql?view=sql-server-ver15) and [sys.dm_exec_connections](https://docs.microsoft.com/sql/relational-databases/system-dynamic-management-views/sys-dm-exec-connections-transact-sql?view=sql-server-ver15).

Please check [the online doc](https://docs.microsoft.com/sql/connect/php/connection-pooling-microsoft-drivers-for-php-for-sql-server?view=sql-server-ver15#linux-and-macos) for details.

In short, you can write a simple php test, like this:
1. Connect and check `connection_id` (call function `getConnectionID` shown below)
2. unset this connection
3. Connect again and check `connection_id` one more time
4. the `connection_id` should be the same if connection pooling is enabled

```
function getConnectionID($conn)
{
    $tsql = "SELECT [connection_id] FROM [sys].[dm_exec_connections] where session_id = @@SPID";
    $stmt = $conn->query($tsql);
    $connID = $stmt->fetchColumn(0);
    unset($stmt);
    return ($connID);
}
```

**Comment by dadamssg** [comment]
thanks @yitam . Unfortunately it doesn't look like i have have permission to do that:

![image](https://user-images.githubusercontent.com/881986/81585312-99454100-9379-11ea-8c72-76bce79c0e2f.png)

**Comment by yitam** [comment]
That I can't help @dadamssg . Please contact your admin to grant you [permission](https://docs.microsoft.com/sql/relational-databases/system-dynamic-management-views/sys-dm-exec-connections-transact-sql?view=sql-server-ver15#permissions), perhaps temporarily, or ask someone who has the permission to check?
